### PR TITLE
Implement ip2nation and ip2nationCountries

### PR DESCRIPTION
**Pull request by Nefertumm**
**Changes proposed:**

- Implemented ip2nation and ip2nationCountries tables. Imported from TrinityCore/TrinityCore@7542049eba46205572b9c9498d6891df31d0ca5b

**Target branch(es):** 1.x
**Issues addressed:** Updates #236 
**Tests performed:** None. If someones can test, i don't have much time.

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1498796247453520600.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO version_db_world (`sql_rev`) VALUES ('1498796247453520600');
+
+DROP TABLE IF EXISTS ip2nation;
+DROP TABLE IF EXISTS ip2nationCountries;
+
+DELETE FROM `command` WHERE `name` in ('account lock', 'account lock ip', 'account lock country');
+INSERT INTO `command` (`name`,`security`,`help`) VALUES
+('account lock ip', 0, 'Syntax: .account lock ip [on|off]\nAllow login from account only from current used IP or remove this requirement.'),
+('account lock country', 0, 'Syntax: .account lock country [on|off]\nAllow login from account only from current used Country or remove this requirement.');
```
Comment (ShinDarth) [inline data/sql/updates/pending_db_world/rev_1498796247453520600.sql]: please add an empty line at the end of file

File: data/sql/updates/pending_db_auth/rev_1498796201292521600.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  code varchar(4) NOT NULL default '',
+  iso_code_2 varchar(2) NOT NULL default '',
+  iso_code_3 varchar(3) default '',
+  iso_country varchar(255) NOT NULL default '',
+  country varchar(255) NOT NULL default '',
+  lat float NOT NULL default '0',
+  lon float NOT NULL default '0',  
+  PRIMARY KEY  (code),
+  KEY code (code)
+);
```
Comment (ShinDarth) [inline data/sql/updates/pending_db_auth/rev_1498796201292521600.sql]: please add an empty line at the end of file

File: modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.cpp [hunk 3]
```diff
@@ -28,6 +28,7 @@ void LoginDatabaseConnection::DoPrepareStatements()
     PrepareStatement(LOGIN_UPD_VS, "UPDATE account SET v = ?, s = ? WHERE username = ?", CONNECTION_ASYNC);
     PrepareStatement(LOGIN_UPD_LOGONPROOF, "UPDATE account SET sessionkey = ?, last_ip = ?, last_login = NOW(), locale = ?, failed_logins = 0, os = ? WHERE username = ?", CONNECTION_SYNCH);
     PrepareStatement(LOGIN_SEL_LOGONCHALLENGE, "SELECT a.sha_pass_hash, a.id, a.locked, a.last_ip, aa.gmlevel, a.v, a.s FROM account a LEFT JOIN account_access aa ON (a.id = aa.id) WHERE a.username = ?", CONNECTION_SYNCH);
+	PrepareStatement(LOGIN_SEL_LOGON_COUNTRY, "SELECT country FROM ip2nation WHERE ip < ? ORDER BY ip DESC LIMIT 0,1", CONNECTION_SYNCH);
```
Comment (ShinDarth) [inline modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.cpp]: please fix indentation

File: modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.cpp [hunk 4]
```diff
@@ -49,6 +50,7 @@ void LoginDatabaseConnection::DoPrepareStatements()
     PrepareStatement(LOGIN_INS_REALM_CHARACTERS_INIT, "INSERT INTO realmcharacters (realmid, acctid, numchars) SELECT realmlist.id, account.id, 0 FROM realmlist, account LEFT JOIN realmcharacters ON acctid=account.id WHERE acctid IS NULL", CONNECTION_ASYNC);
     PrepareStatement(LOGIN_UPD_EXPANSION, "UPDATE account SET expansion = ? WHERE id = ?", CONNECTION_ASYNC);
     PrepareStatement(LOGIN_UPD_ACCOUNT_LOCK, "UPDATE account SET locked = ? WHERE id = ?", CONNECTION_ASYNC);
+	PrepareStatement(LOGIN_UPD_ACCOUNT_LOCK_CONTRY, "UPDATE account SET lock_country = ? WHERE id = ?", CONNECTION_ASYNC);
```
Comment (ShinDarth) [inline modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.cpp]: please fix indentation

File: modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.cpp [hunk 5]
```diff
@@ -76,5 +78,6 @@ void LoginDatabaseConnection::DoPrepareStatements()
     PrepareStatement(LOGIN_SEL_ACCOUNT_WHOIS, "SELECT username, email, last_ip FROM account WHERE id = ?", CONNECTION_SYNCH);
     PrepareStatement(LOGIN_SEL_REALMLIST_SECURITY_LEVEL, "SELECT allowedSecurityLevel from realmlist WHERE id = ?", CONNECTION_SYNCH);
     PrepareStatement(LOGIN_DEL_ACCOUNT, "DELETE FROM account WHERE id = ?", CONNECTION_ASYNC);
+	PrepareStatement(LOGIN_SEL_IP2NATION_COUNTRY, "SELECT c.country FROM ip2nationCountries c, ip2nation i WHERE i.ip < ? AND c.code = i.country ORDER BY i.ip DESC LIMIT 0,1", CONNECTION_SYNCH);
```
Comment (ShinDarth) [inline modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.cpp]: please fix indentation

File: modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.h [hunk 6]
```diff
@@ -46,6 +46,7 @@ enum LoginDatabaseStatements
     LOGIN_UPD_VS,
     LOGIN_UPD_LOGONPROOF,
     LOGIN_SEL_LOGONCHALLENGE,
+	LOGIN_SEL_LOGON_COUNTRY,
```
Comment (ShinDarth) [inline modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.h]: please fix indentation

File: modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.h [hunk 7]
```diff
@@ -69,6 +70,7 @@ enum LoginDatabaseStatements
     LOGIN_INS_REALM_CHARACTERS_INIT,
     LOGIN_UPD_EXPANSION,
     LOGIN_UPD_ACCOUNT_LOCK,
+	LOGIN_UPD_ACCOUNT_LOCK_CONTRY,
```
Comment (ShinDarth) [inline modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.h]: please fix indentation

File: modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.h [hunk 8]
```diff
@@ -96,6 +98,7 @@ enum LoginDatabaseStatements
     LOGIN_SEL_ACCOUNT_WHOIS,
     LOGIN_SEL_REALMLIST_SECURITY_LEVEL,
     LOGIN_DEL_ACCOUNT,
+	LOGIN_SEL_IP2NATION_COUNTRY,
```
Comment (ShinDarth) [inline modules/worldengine/nucleus/src/Database/Implementation/LoginDatabase.h]: please fix indentation

File: src/authserver/Server/AuthSocket.cpp [hunk 9]
```diff
@@ -417,8 +417,36 @@ bool AuthSocket::_HandleLogonChallenge()
                 else
                     ;//sLog->outDebug(LOG_FILTER_NETWORKIO, "[AuthChallenge] Account IP matches");
             }
-            else
+			else
```
Comment (ShinDarth) [inline src/authserver/Server/AuthSocket.cpp]: please fix indentation

File: src/authserver/Server/AuthSocket.cpp [hunk 10]
```diff
@@ -417,8 +417,36 @@ bool AuthSocket::_HandleLogonChallenge()
...
+							pkt << uint8(WOW_FAIL_UNLOCKABLE_LOCK);
+							locked = true;
+						}
+						else
+							;//sLog->outDebug(LOG_FILTER_NETWORKIO, "[AuthChallenge] Account country matches");
+					}
+                    else
+                        ;//sLog->outDebug(LOG_FILTER_NETWORKIO, "[AuthChallenge] IP2NATION Table empty");
+                }
+            }
```
Comment (ShinDarth) [inline src/authserver/Server/AuthSocket.cpp]: please fix indentation of all this blocks....

File: src/scripts/Commands/cs_account.cpp [hunk 11]
```diff
@@ -230,7 +235,59 @@ class account_commandscript : public CommandScript
...
+				LoginDatabase.Execute(stmt);
+                handler->PSendSysMessage(LANG_COMMAND_ACCLOCKUNLOCKED);
+            }
+            return true;
+        }
+
+        handler->SendSysMessage(LANG_USE_BOL);
+        handler->SetSentErrorMessage(true);
+        return false;
+    }
```
Comment (ShinDarth) [inline src/scripts/Commands/cs_account.cpp]: indentation is bad here too, please fix it

**Comment by ShinDarth** [comment]
@Nefertumm good job, but please check my comments

**Comment by Nefertumm** [comment]
@ShinDarth done :+1:

**Comment by ShinDarth** [comment]
thanks, waiting for a test feedback

**Comment by Yehonal** [comment]
```
Query:
ALTER TABLE `account` ADD COLUMN `lock_country` VARCHAR(2) NOT NULL DEFAULT '00' AFTER `locked`

Error occured at:2017-07-31 14:20:15
Line no.:3
Error Code: 1067 - Invalid default value for 'last_login'
```

i've this error importing your sql

**Comment by Yehonal** [comment]
@Nefertumm did you tested it? 

Because if you try to run the authserver and login it crashes since you have issues with fields, on various queries i think, such as LOGIN_SEL_ACCOUNT_BANNED, 

crash log:

`Thread 1 "authserver" received signal SIGSEGV, Segmentation fault.
strlen () at ../sysdeps/x86_64/strlen.S:106
1502900766:196962 106	../sysdeps/x86_64/strlen.S: No such file or directory.
#0  strlen () at ../sysdeps/x86_64/strlen.S:106
#1  0x00007ffff6902c37 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x000000000044424c in Field::GetString[abi:cxx11]() const (this=0x77b5bf) at /home/azth/usr/workspace/projects/azerothcore-wotlk/modules/worldengine/nucleus/src/Database/Field.h:236
#3  0x000000000044839f in AuthSocket::_HandleLogonChallenge (this=0x77b8e0) at /home/azth/usr/workspace/projects/azerothcore-wotlk/src/authserver/Server/AuthSocket.cpp:480
#4  0x0000000000446df7 in AuthSocket::OnRead (this=0x77b8e0) at /home/azth/usr/workspace/projects/azerothcore-wotlk/src/authserver/Server/AuthSocket.cpp:234
#5  0x0000000000454369 in RealmSocket::handle_input (this=0x77b740) at /home/azth/usr/workspace/projects/azerothcore-wotlk/src/authserver/Server/RealmSocket.cpp:266
#6  0x00007ffff6c388cc in ACE_Dev_Poll_Reactor::dispatch_io_event(ACE_Dev_Poll_Reactor::Token_Guard&) () from /usr/lib/libACE-6.3.3.so
#7  0x00007ffff6c3a46e in ACE_Dev_Poll_Reactor::handle_events(ACE_Time_Value*) () from /usr/lib/libACE-6.3.3.so
#8  0x00007ffff6c7b382 in ACE_Reactor::run_reactor_event_loop(ACE_Time_Value&, int (*)(ACE_Reactor*)) () from /usr/lib/libACE-6.3.3.so
#9  0x0000000000459187 in main (argc=3, argv=0x7fffffffe0e8) at /home/azth/usr/workspace/projects/azerothcore-wotlk/src/authserver/Main.cpp:282
`
the issue is on Server/AuthSocket.cpp:480

where fields[6] and fields[7] are not related with LOGIN_SEL_ACCOUNT_BANNED

**Comment by Yehonal** [comment]
fixed and tested

needs minor improvements on WorldSocket.cpp but it's not mandatory for now

Merged

**Pull request merged**